Revise - WebEngine support for SDL JavaScript (#782)
smartdevicelink/sdl_evolution: proposals/0240-sdl-js-pwa.md

App registration on the SDL Developer Portal (smartdevicelink.com) should allow a developer to specify an app as a local web app. As the app platforms increase (2 -> 3) a new way to specify preferences should be introduced. The element `BOTH` should be deprecated and replaced by `ALL` for a next major release. Occurrences of `BOTH` would be treated as `ALL`.

### 1.4 HMI API using App Properties RPCs

The HMI API should be extended to set app properties to SDL Core. This addition should allow the HMI to add app policy IDs to the policy database as they are installed by the embedded OEM catalog.

The HMI API extension is mostly a copy of the cloud-app-properties included in the mobile API. 

1. The OEM store uses `SetAppProperties` using the manifest data.

> 2. The OEM store may choose add not-installed apps to SDL using `enabled` parameter set to `false` (optional)

Reviewer: ```suggestion
2. The OEM store may choose to add not-installed apps to SDL using `enabled` parameter set to `false` (optional)
```

##### 2.2.3 User interface guidelines (Driver Distraction rules)

With the current depth of this chapter, the new HMI type should be used by 1st party OEM apps only. With future proposals and workshops the SDLC should open the new HMI type to 3rd party by creating and defining proper driver distraction and user interface guidelines.

At the time of this proposal being in review, a set of driver distraction rules are being created and proposed to enable 3rd party using the open HMI. The following bullet points list items that will be described further in the ruleset:

-	minimum font size
-	minimum contrast between font and background
-	min/max brightness (for day and night mode)
-	number of buttons
-	minimum size of a button
- No customized keyboard
- No video playback (exceptions in standstill per countries)
-	NHTSA related guidelines
  o	Amount of text (button and text fields)
  o	number of lines of text
  o	Complexity of use cases (number of steps to fulfill a use case)

More items may be included in the ruleset as they become Driver Distraction affected.

##### 2.2.4 System context and event change

> Independent of the app presentation type, the HMI will continue provide system context information from the app. An application which uses the open HMI should continue to recieve `OnHMIStatus` notifications and SDL Core will still be notified about event changes.

Reviewer: ```suggestion
Independent of the app presentation type, the HMI will continue to provide system context information from the app. An application which uses the open HMI should continue to recieve `OnHMIStatus` notifications and SDL Core will still be notified about event changes.
```

## Alternatives considered

Local Node.js or Java were considered as alternative options for locally running applications. However both options have downsides:
1. Both are difficult to sandbox. Compared to a web engine the effort to sandbox a Node.js or java application and to protect the vehicle system are very high.

> 2. Limited app availability. App developers would potentially need to reassemble a godd portion of existing code and write new code to make locally running applications possible.

Reviewer: ```suggestion
2. Limited app availability. App developers would potentially need to reassemble a good portion of existing code and write new code to make locally running applications possible.
```

3. Licensing and compatibility. License cost may apply for embedded in-vehicle use. Efforts avoiding license using older versions may cause compatiblity issues leading to code rewrite. Open source variants may cause other license issues.

> Many services are available over a web appliation and modern WebEngines provide a good sandboxing ability. Last the ability of an open HMI is outstanding and not supported by either of the alternatives.

Reviewer: ```suggestion
Many services are available over a web application and modern WebEngines provide a good sandboxing ability. Lastly, the ability of an open HMI is outstanding and not supported by either of the alternatives.
```
